docs: introduce KeptnApp to KeptnAppContext migration guide (#2851)
keptn/lifecycle-toolkit: docs/docs/migrate/keptnapp/index.md

# Migrating KeptnApp to KeptnAppContext

> Keptn has recently orphaned the concept of manually created KeptnApp
> and instead uses the KeptnApp auto-discovery feature combined with
> the newly introduced KeptnAppContext custom resource.

Reviewer: ```suggestion
The introduction of the `KeptnAppContext` resource
in Keptn v0.11.0
requires modifications to existing
[KeptnApp](../reference/crd-reference/app)
resources that were generated manually.
This page gives instructions for doing that.
```

> ## Migration steps
> 
> > **Warning**
> Please note that the migration is needed to be executed only for
> **manually created KeptnApp** resources.
> KeptnApp resources created by
> the automatic app-discovery feature do not need any actions.

Reviewer: ```suggestion

> **Note**
Migration is only required for
**manually created KeptnApp** resources.
KeptnApp resources created by
the
[automatic app-discovery](../guides/auto-app-discovery)
feature do not require modification
unless you edited them manually to add pre/post tasks and evaluations.

## Migration steps
```
* I don't think this warrants a "Warning" -- that should be reserved for actions that have dire consequences
* The admonition should come before the section that gives how-to info
* I added the "unless" but I don't know if that is true.  But we need to address the hybrid approach that we've been recommending, to auto-generate KeptnApp then edit it to add tasks and evaluations

Author: I would stick to warning here, as it's very important that the users know that they need to migrate only if the KeptnApp was created manually. Just for visibility

To migrate an existing **manually created KeptnApp** to KeptnAppContext
you need to execute the following steps:

> 1. Create a KeptnAppContext custom resource with the same name
> as your KeptnApp has

Reviewer: ```suggestion
1. Create a KeptnAppContext custom resource that has the same name
as your KeptnApp.
```

> 2. Move pre/post-deployment tasks and evaluation from KeptnApp
> to KeptnAppContext

Reviewer: ```suggestion
2. Move the lists of pre/post-deployment tasks and evaluation from `KeptnApp`
to `KeptnAppContext`.
In other words, delete them from `KeptnApp` and add them to `KeptnAppContact`.
```

> 3. Introduce a `app.kubernetes.io/managed-by: keptn` annotation

Reviewer: ```suggestion
3. Add the `app.kubernetes.io/managed-by: keptn` annotation
```

> to KeptnApp

Reviewer: ```suggestion
to `KeptnApp`
```

> 4. Change the API version of KeptnApp to `lifecycle.keptn.sh/v1beta1`

Reviewer: ```suggestion
4. Change the value of the `apiVersion` field for the `KeptnApp` resource to `lifecycle.keptn.sh/v1beta1`
```

> **Note**

> Please make sure all of your application resources

Reviewer 2: Here we can add a note that for workloads that have been configured correctly previously, no action will be needed for the migration, as the `part-of`/`app` labels/annotations will be set properly (otherwise the deployment of the KeptnApp and the workloads would not have worked anyway)

(such as Pods, Deployments, StatefulSets or DaemonSets)

> have the proper annotations/labels set.

Reviewer: So do you men that all resources listed in the `KeptnApp` `spec.workoads` require these annotations?  I think it would be clearer to reference them this way rather than as "application resources", which is not the terminology that is used elsewhere.

Author: not particulary the resources listed in spex.workloads, but the resources, which are represented by the workloads (Deployments, StatefulSets, Pods...)

Reviewer: Okay, when we get everything updated, we can look to be sure that our terminology is consistent.  It may be that some of the existing docmentation needs to switch terminology.  I am resolving the comment.

These annotations/labels (especially the
`app.kubernetes.io/part-of` or `keptn.sh/app`)
are needed for the full migration to the
usage of automatic app-discovery feature.
More information about how to set up these annotations/labels
can be found [here](../../guides/integrate.md#basic-annotations).

## Example of migration

> In the next subsection we are going to look at an example of the migration.

Reviewer: ```suggestion
In the next subsection we are going to look at an example of how to modify the `KeptnApp` definition.
```
I'm thinking that this is not really the full migration process, just how to modify this one resource definition.

Let's say we have the following KeptnApp in our cluster:

```yaml
{% include "./asserts/keptnapp.yaml" %}
```

> If we apply the migration steps from the previous subsection, we will ge the

Reviewer: ```suggestion
If we apply the migration steps from the previous subsection, we get the
```

following result:

```yaml
{% include "./asserts/keptnapp-migrated.yaml" %}
```

These modified resources can be now applied to your cluster.

> **Note**

> Please make sure all of your application resources

Reviewer: Same question as above about calling these "application resources"

(such as Pods, Deployments, StatefulSets or DaemonSets)
have the proper annotations/labels set.
More information about how to set up these annotations/labels

> can be found [here](../../guides/integrate.md#basic-annotations).

Reviewer: Do we really need to duplicate this information here.  It seems that twice in such a short document is overkill but you may not agree.

Author: it's not needed, we can remove it

> # Migrating KeptnApp to KeptnAppContext

Reviewer: @thisthat Yup.  Updating the KeptnApp ref page and creating a KeptnAppContext reference page are in a separate PR that I am doing.

The introduction of the `KeptnAppContext` resource
in Keptn v0.11.0
requires modifications to existing
[KeptnApp](../../reference/crd-reference/app.md)
resources that were generated manually.
This page gives instructions for doing that.

## Migration steps

> **Warning**
Migration is only required for
**manually created KeptnApp** resources.
KeptnApp resources created by
the
[automatic app-discovery](../../guides/auto-app-discovery.md)
feature do not require modification

> unless you edited them manually to add pre/post tasks and evaluations.

Author: it says in the warning:

Migration is only required for
**manually created KeptnApp** resources.
`KeptnApp` resources created by Keptn via the
automatic app-discovery
feature do not require modification
unless you edited them manually to add pre/post-deployment tasks or evaluations.

It says it quite clearly in my opinion: if you created the KeptnApp manually or you edited the auto generated one, here are actions you need to take...

> unless you edited them manually to add pre/post tasks and evaluations.

Author: adapted

> To migrate an existing **manually created KeptnApp** to KeptnAppContext

Author: That's not possible in the previous configuration, unless the user didn't edit it. And that we have in the previous paragraph:

...unless you edited them manually to add pre/post tasks and evaluations.

> To migrate an existing **manually created KeptnApp** to KeptnAppContext

Author: I adapted it to be more clear

# Migrating KeptnApp to KeptnAppContext

The introduction of the `KeptnAppContext` resource
in Keptn v0.11.0
requires modifications to existing
[KeptnApp](../../reference/crd-reference/app.md)
resources that were generated manually.
This page gives instructions for doing that.

## Migration steps

> **Warning**

> Migration is only required for

Reviewer: I would prefer to have the admonition before the header.  If you disagree, just resolve this comment.

# Migrating KeptnApp to KeptnAppContext

The introduction of the `KeptnAppContext` resource
in Keptn v0.11.0
requires modifications to existing
[KeptnApp](../../reference/crd-reference/app.md)
resources that were generated manually.
This page gives instructions for doing that.

## Migration steps

> **Warning**
Migration is only required for
**manually created KeptnApp** resources.
KeptnApp resources created by
the
[automatic app-discovery](../../guides/auto-app-discovery.md)
feature do not require modification
unless you edited them manually to add pre/post tasks and evaluations.

To migrate an existing **manually created KeptnApp** to KeptnAppContext
you need to execute the following steps:

> 1. Create a KeptnAppContext custom resource that has the same name

Reviewer: ```suggestion
1. Create a `KeptnAppContext` custom resource that has the same name
```

> as your KeptnApp.

Reviewer: ```suggestion
as your `KeptnApp`.
```
